src/main/scala/mesosphere/marathon/core/health/impl/AppHealthCheckActor.scala

```diff
@@ -0,0 +1,157 @@
+package mesosphere.marathon
+package core.health.impl
+
+import akka.actor.{ Actor, Props }
+import akka.event.EventStream
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.core.event.InstanceHealthChanged
+import mesosphere.marathon.core.health.impl.AppHealthCheckActor._
+import mesosphere.marathon.core.health.{ Health, HealthCheck }
+import mesosphere.marathon.core.instance.Instance
+import mesosphere.marathon.state.{ PathId, Timestamp }
+
+import scala.collection.mutable
+
+object AppHealthCheckActor {
+  case class ApplicationKey(appId: PathId, version: Timestamp)
+  case class InstanceKey(applicationKey: ApplicationKey, instanceId: Instance.Id)
+
+  def props(eventBus: EventStream): Props = Props(new AppHealthCheckActor(eventBus))
+
+  case class AddHealthCheck(appKey: ApplicationKey, healthCheck: HealthCheck)
+  case class RemoveHealthCheck(appKey: ApplicationKey, healthCheck: HealthCheck)
+
+  case class PurgeHealthCheckStatuses(hc: Seq[(InstanceKey, HealthCheck)])
+  case class HealthCheckStatusChanged(
+      appKey: ApplicationKey,
+      healthCheck: HealthCheck, health: Health)
+
+  trait InstanceHealthChangedNotifier {
+    def notify(applicationKey: ApplicationKey, instanceId: Instance.Id,
+      healthiness: Option[Boolean]): Unit
+  }
+
+  class AppHealthCheckProxy extends StrictLogging {
+    /**
+      * Map of health check definitions of all applications
+      */
+    var healthChecks: mutable.Map[ApplicationKey, Set[HealthCheck]] = mutable.Map.empty
```

Comment:
Mutable variables should be `private [this]`

Reply:
Also, double mutable (var + mutable datastructure). Prefer:

```
val healthChecks: mutable.Map[...]
```

or

```
var healthChecks: immutable.Map[...]
```

But not both :-)

Reply:
I changed `var` to `val` but I cannot make the variables private otherwise we will lose a part of the test coverage. Some tests are checking if the purge works as expected.

Reply:
Well, I eventually made them private and end up creating methods to check existence of objects in the maps so that I can use them in the tests.

Reply:
You can make something private such that the test can access is, `private [impl]` (package scoped privates)

Reply:
Ok, I'm trying now. Can you tell me what's wrong with the build in the meantime?

Reply:
done. Tests are green on my side.